Change object storage cluster table functions to prefer specific replicas to improve cache locality

adikus opened the pull request:
### Changelog category (leave one):
- Improvement

### Changelog entry (a user-readable short description of the changes that goes to CHANGELOG.md):
Object storage cluster table functions (e.g. `s3Cluster`) will now assign files to replicas for reading based on consistent hash to improve cache locality.

### Documentation entry for user-facing changes

- [ ] Documentation is written (mandatory for new features)

<!---
Directly edit documentation source files in the "docs" folder with the same pull-request as code changes

or

Add a user-readable short description of the changes that should be added to docs.clickhouse.com below.

At a minimum, the following information should be added (but add more as needed).
- Motivation: Why is this function, table engine, etc. useful to ClickHouse users?

- Parameters: If the feature being added takes arguments, options or is influenced by settings, please list them below with a brief explanation.

- Example use: A query or command.
-->

Addresses: #72816
This works by assigning files from the file iterator to specific replicas. 
This PR also implements task stealing, so that if a replica is not available or is slow, other replicas will process its tasks. In terms of filesystem cache locality, though, this means that at the end of iterating through tasks, some task stealing is inevitable, preventing us from achieving perfect cache locality. ~~I've added a 50ms sleep before the stealing is allowed to proceed, which helps a bit, but I'm not sure if that's a good approach.~~

~~Added a `object_storage_stable_cluster_task_distribution` setting to enable the new behaviour.~~

I'm happy to write some tests if I get a 👍 on the approach.

adikus committed:
Change object storage cluster table functions to prefer specific replicas to improve cache locality

nickitat commented on src/Storages/ObjectStorage/StorageObjectStorageStableTaskDistributor.cpp:
```diff
@@ -0,0 +1,223 @@
+    if (total_replicas <= 0)
+        return 0;
+        
+    UInt64 hash_value = sipHash64(file_path);
```
pls use consistentHash

nickitat commented on src/Storages/ObjectStorage/StorageObjectStorageStableTaskDistributor.cpp:
```diff
@@ -0,0 +1,223 @@
+        return next_file;
+    }
+    
+    return "";
```
pls use std::optional instead of plain String

nickitat commented on src/Storages/ObjectStorage/StorageObjectStorageStableTaskDistributor.cpp:
```diff
@@ -0,0 +1,223 @@
+    return static_cast<Int32>(hash_value % total_replicas);
+}
+
+String StorageObjectStorageStableTaskDistributor::getPreQueuedFile(const String & connection_key)
```
if we will pass a replica_num, connection id will be a normal integer

nickitat commented on src/Storages/ObjectStorage/StorageObjectStorageStableTaskDistributor.cpp:
```diff
@@ -0,0 +1,223 @@
+    if (!file.empty())
+        return file;
+
+    if (!unprocessed_files.empty()) {
+        // Prevent initiator from stealing jobs from other replicas 
+        sleepForMilliseconds(50);
+    }
```
let's not do it

adikus commented:
Thanks for the review @nickitat, planning to address your comments either today or tomorrow

adikus committed:
Rewrite cluster function stable mapping to use number_of_current_replica

adikus committed:
Rewrite cluster function stable mapping to use number_of_current_replica :lipstick:

adikus committed:
Use consistent hash in StorageObjectStorageStableTaskDistributor

adikus committed:
Use std::optional<String> in StorageObjectStorageStableTaskDistributor

adikus commented on src/Storages/IStorageCluster.cpp:
```diff
@@ -183,31 +183,46 @@ void ReadFromCluster::initializePipeline(QueryPipelineBuilder & pipeline, const
+    {
+        IConnections::ReplicaInfo replica_info{
+            .number_of_current_replica = replica_index++,
+            .number_of_replicas = number_of_replicas,
```
Not sure sure about this. I've seen that for parallel replicas we hold `replicas_count` inside the `ParallelReplicasReadingCoordinator`, not sure if it makes sense to use that here.

adikus committed:
Merge branch 'master' into ah/stable-s3-filesystem-cache

# Conflicts:
#	src/Storages/IStorageCluster.cpp
#	src/Storages/ObjectStorage/StorageObjectStorageCluster.cpp

adikus committed:
Use getPathToArchive in StorageObjectStorageStableTaskDistributor as well

adikus committed:
Fix trailing whitespaces in StorageObjectStorageStableTaskDistributor

adikus committed:
Add object_storage_stable_cluster_task_distribution to settings changes history

adikus commented:
@nickitat Rewritten to use `number_of_current_replica`, let me know what you think. 🙂

alexey-milovidov commented:
Thanks for the amazing work, it looks like exactly what we need!
Let's take a look at failed tests.

adikus committed:
Remove object_storage_stable_cluster_task_distribution setting

adikus committed:
Address review comments
Initialize task distributor with number of replicas directly, instead of passing it through replica info

adikus committed:
Merge branch 'master' into ah/stable-s3-filesystem-cache

adikus committed:
Remove object_storage_stable_cluster_task_distribution setting :white_check_mark:

adikus committed:
Address review comments :white_check_mark:

adikus commented:
Thanks @nickitat and sorry for the delay here.
I incorporated your feedback, let me know how it looks.

adikus commented on src/Storages/IStorageCluster.cpp:
```diff
@@ -104,15 +104,19 @@ void ReadFromCluster::applyFilters(ActionDAGNodes added_filter_nodes)
         predicate = filter_actions_dag->getOutputs().at(0);
 
-    createExtension(predicate);
+    auto max_replicas_to_use = static_cast<UInt64>(cluster->getShardsInfo().size());
+    if (context->getSettingsRef()[Setting::max_parallel_replicas] > 1)
+        max_replicas_to_use = std::min(max_replicas_to_use, context->getSettingsRef()[Setting::max_parallel_replicas].value);
+
+    createExtension(predicate, max_replicas_to_use);
```
One thing I wasn't quite sure is how the the initialization works here in the `applyFilters` - not sure what is the code path to get here.

I copied parts from below to get the same number of replicas, so hopefully that will work ok.

adikus committed:
Address review comments :white_check_mark:

adikus committed:
Address review comments :white_check_mark:

adikus committed:
Address review comments :white_check_mark:

adikus commented:
Looks like two tests failed on the "Replica info is not initialized" checks, will have a look at those tomorrow.

adikus committed:
Properly initialize replica info wherever we use task iterator

adikus commented:
Initialized replica info on all extensions that also use task iterator, even though it is not currently used for most of them. Not sure if this is going too deep in this PR.

adikus committed:
Merge branch 'master' into ah/stable-s3-filesystem-cache

# Conflicts:
#	src/Storages/IStorageCluster.cpp
#	src/Storages/StorageDistributed.cpp
